### DISCUSSION TITLE
Features to be deprecated in 7.1

### DISCUSSION THREAD

**Discussion by ericjbohm** [UIUC-PPL/charm#3682] [Development]
We intend to clean up older unused features and defunct machine layers in the next major release.

**Comment by lvkale** [comment]
Please list potential features so we can discuss them.
In the group meeting we had identified: 
machine layers: uGNI, possibly PAMI and some others.
message-logging based scalable checkpoint/restart.

**Reply by ericjbohm** [reply]
POSE and related bigsim infrastructure

**Reply by ericjbohm** [reply]
Multiblock and AMR libraries

**Reply by ericjbohm** [reply]
PARFUM and related libraries

**Reply by ericjbohm** [reply]
Explore CharmXI for defunct features.

**Reply by rbuch** [reply]
The Tau tracing module?

**Reply by stwhite91** [reply]
> Explore CharmXI for defunct features.

The "atomic" keyword has been deprecated in favor of serial since v6.5.0 according to the release notes, but the manual still only says we are considering deprecation of it. I think atomic should be removed eventually, but it appears to still be used in completion detection and some examples.

**Reply by evan-charmworks** [reply]
- mlog and causal ft and bigsim have already been removed. POSE is still around; I didn't realize it depended on BigSim?
- We removed FEM because ParFUM replaces it, so why remove ParFUM now?
- The PAMI layer is a strong candidate for removal, however the performance gap between it and PAMILRTS should be measured and rectified: #2040
- uGNI support should go if there are no Cray XC or XE machines around any more, and if so, there is other code that could go too such as the MPI Cray XC/XE layers, and other Cray specific code that can be found by grepping. Blue Waters is gone, and Cori is about to be decommissioned, but I don't know if the platform is completely gone yet like Blue Gene.

**Comment by stwhite91** [comment]
I think ParFUM, ParFUM-Iterators, ParFUM-Tops, ParFUM-Tops-Dev, idxl, pose, mblock, amr, irecv, and barrier can all be removed from ck-libs, as well as master-slave and lrpc from conv-libs. Maybe others, but I'm not sure-- perhaps Core should go through the libs/ directories to identify which ones we want removed.

**Comment by evan-charmworks** [comment]
Here are criteria I would use for deciding whether to remove something. If the answer to any of these is yes, it is worth considering removal. If they are all no, there is probably no harm in keeping it. For each, I give examples of things we have removed in my time working with Charm.

- Is the code for a dead target? (ex: BlueGene support, PlayStation 3/Cell/SPE charmxi feature, bproc/clustermatic/scyld)
- Has the component been supplanted by something else? (ex: FEM replaced by ParFUM, net layer --> netlrts, CmiDirect API --> Zero Copy API)
- Is there a more than minimal maintenance burden? (ex: BigSim, AMPI multiple worlds, message logging fault tolerance)
- Is the implementation a stub? (ex: SHMEM machine layer)

IMO, "no one is using it" is not a strong justification on its own. I have experience from other projects where things were removed only due to this reason, and I had to revert the removal because it ended up being incorrect!
